test/extended/setup_compat.sh

```diff
@@ -0,0 +1,354 @@
+#!/bin/bash
+#
+# Starts an OpenShift master for compatibility testing
+
+
+# Launches a server for OpenShift compatibility testing
+# PREVIOUS_CONTROLLER=1 launches with previous version controllers
+# PREVIOUS_API=1        launches with previous version API server
+
+# Sets up a current master server with previous controllers
+function os::test::extended::setup_compatibility {
+  # os::log::stacktrace::install
+
+  # build binaries
+  if [[ -z $(os::build::find-binary ginkgo) ]]; then
+    hack/build-go.sh vendor/github.com/onsi/ginkgo/ginkgo
+  fi
+  if [[ -z $(os::build::find-binary extended.test) ]]; then
+    hack/build-go.sh test/extended/extended.test
+  fi
+  if [[ -z $(os::build::find-binary openshift) ]]; then
+    hack/build-go.sh
+  fi
+  if [[ -z "${PREVIOUS_VERSION:-}" ]]; then
+    if [[ -z $(os::build::find-binary previousversion) ]]; then
+      hack/build-go.sh tools/previousversion
+    fi
+    openshift_binary=$(os::build::find-binary openshift)
+    previous_version=$(os::build::find-binary previousversion)
+    export PREVIOUS_VERSION="$(${previous_version} ${openshift_binary})"
+  fi
+  if [[ -z "${API_SERVER_VERSION:-}" ]] && [[ -n "${PREVIOUS_API:-}" ]]; then
+    API_SERVER_VERSION="${PREVIOUS_VERSION}"
+  fi
+  if [[ -z "${CONTROLLER_VERSION:-}" ]] && [[ -n "${PREVIOUS_CONTROLLER:-}" ]]; then
+    CONTROLLER_VERSION="${PREVIOUS_VERSION}"
+  fi
+
+  os::util::environment::setup_time_vars
+
+  # ensure proper relative directories are set
+  export GINKGO="$(os::build::find-binary ginkgo)"
+  export EXTENDEDTEST="$(os::build::find-binary extended.test)"
+  export TMPDIR=${BASETMPDIR:-/tmp}
+  export EXTENDED_TEST_PATH="$(pwd)/test/extended"
+  export KUBE_REPO_ROOT="$(pwd)/vendor/k8s.io/kubernetes"
+
+  # output tests instead of running
+  if [[ -n "${SHOW_ALL:-}" ]]; then
+    TEST_OUTPUT_QUIET=true ${EXTENDEDTEST} --ginkgo.dryRun --ginkgo.noColor | grep ok | grep -v skip | cut -c 20- | sort
```

Comment:
Nobody writes or understands code like that, though... instead of checking at every step that we can *still* move forward, just write the code as if it were not Bash. I really don't get the aversion to the Bash `for` loop -- sure, there *are* constructs in Bash that, while idiomatic, arguably are hard to understand and maybe should be avoided. I believe the long tail of pipes is such a construct, that, while useful on the command line for a simple action, does not belong in an enterprise script that needs to be read, understood, and maintained by a varied group of people over a long period of time.

The largest issue I have with the latest proposal is that it is difficult to know exactly *what* your `if` blocks are checking -- without a nuanced understanding, it is not immediately clear if they will trigger when the command substitution inside of `$()` fails, or if the variable assignment fails, or both, or either. I don't want someone reading this to need that nuanced understanding.

With Bash idioms, your code looks like:
```sh
if [[ -n "${SHOW_ALL:-}" ]]; then
	test_list="$( "${EXTENDEDTEST}" --ginkgo.dryRun --ginkgo.noColor )"
	if grep -q "ok" <<< "${test_list[@]}"; then
		test_list="$( grep "ok" <<< "${test_list[@]}" )"
		if grep -qv "skip" <<< "${test_list[@]}"; then
			test_list="$( grep -v "skip" <<< "${test_list[@]}" )"
			sort <<< "${test_list[@]}"
		fi
	fi
fi
```

For comparison, here is my original suggestion:
```sh
if [[ -n "${SHOW_ALL:-}" ]]; then
	test_list="$( "${EXTENDEDTEST}" --ginkgo.dryRun --ginkgo.noColor )"
	for test in "${test_list[@]}"; do
		if grep -q "ok" <<< "${test}" && grep -qv "skip" <<< "${test}"; then
			echo "${test}"
		fi
	done
fi
```

If we deem the sorting to really be critical...
```sh
if [[ -n "${SHOW_ALL:-}" ]]; then
	full_test_list="$( "${EXTENDEDTEST}" --ginkgo.dryRun --ginkgo.noColor )"
	selected_tests=()
	for test in "${full_test_list[@]}"; do
		if grep -q "ok" <<< "${test}" && grep -qv "skip" <<< "${test}"; then
			selected_tests+=( "${test}" )
		fi
	done
	sort <<< "${selected_tests[@]}"
fi
```

I will continue to argue for this form as it is much more readable, flows like a normal program, and does not appear to me to have subtleties to it.